src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,137 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    public static final int DEFAULT_CAPACITY = 10;
+    private static final Object[] EMPTY_ELEMENT_DATA = {};
+    private Object[] elementData;
+    private int size;
+
+    public ArrayList() {
+        elementData = EMPTY_ELEMENT_DATA;
+        size = 0;
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == elementData.length) {
+            elementData = grow();
+        }
+        elementData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        rangeCheckForAdd(index);
+        int countElementsAfterIndex = size - index;
+        Object[] elementData;
+        if (size == (elementData = this.elementData).length) {
+            elementData = grow();
+        }
+        System.arraycopy(elementData, index,
+                elementData, index + 1,
+                countElementsAfterIndex);
+        elementData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        int amountListElements = list.size();
+        Object [] arrayFromList = toArray(list, amountListElements);
+        Object[] elementData;
+        final int s;
+        if (amountListElements > (elementData = this.elementData).length - (s = size)) {
+            elementData = grow(s + amountListElements);
+        }
+        System.arraycopy(arrayFromList, 0, elementData, s, amountListElements);
+        size = s + amountListElements;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return (T) elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        int countElementsAfterIndex = elementData.length - 1 - index;
+        Object removedElement = elementData[index];
+        System.arraycopy(elementData,index + 1,
+                elementData,index,countElementsAfterIndex);
+        size -= 1;
+        return (T) removedElement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int elementIndex = findElementIndex(element);
+        return remove(elementIndex);
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Object[] grow() {
+        return grow(size + 1);
+    }
+
+    private Object[] grow(int minCapacity) {
```

Comment:
when using this method you always assign this array to this.elements
lets assign it here and make this method return void